doc/rst/language/spec/arrays.rst

```diff
@@ -162,23 +162,42 @@ Rectangular Array Literals
 
 Rectangular array literals are specified by enclosing a
 comma-separated list of expressions in square brackets, where each
-expression represents an array element's value.  A trailing comma is
-permitted after the final array element.  For a literal with `n`
+expression represents an array element's value. A trailing comma is
+permitted after the final array element. For a literal with `n`
 expressions, an anonymous domain literal ``{0..<n}`` is generated to
-represent the array's indices.  If the value expressions are all of
+represent the array's indices. If the value expressions are all of
 the same type, that type will be used as the array's element type.  If
 they are not, the array's element type is computed using the same type
 unification logic that is used when inferring the return type of a
 procedure with an implicit return type (see
 :ref:`Implicit_Return_Types`).
 
-
+A multi-dimensional rectangular array literal is specified using
+comma-separated values, with each dimension separated by a semicolon. Multiple
+semicolons are used to denote the next dimension. An array literal of rank
+``n`` uses ``n-1`` semicolons to separate the last dimension. Only
+rectilinear multi-dimensional literals are supported, the number of elements in
+each dimension must be the same. A single trailing semicolon is permitted after
+the final row and a trailing comma is permitted after the final element in each
+row.
```

Comment:
Is the "single trailing" accurate?  E.g., in a 3D array, can I not have a trailing semicolon per dimension to write:

```
[1, 2;
 3, 4;
  ;
  5, 6;
  7, 8;
  ;
]
```

(if not, we shouldn't hold up this PR over it, but it seems like it might be nice to open an issue for and add as an extension later.  Specifically, if I'm machine-generating literals, it's convenient not to have to special case "is this the last thing in this row/plane/cube/etc.)

Reply:
In this PR, I've explicitly disallowed that. you can either have no trailing semicolon, or a single trailing semicolon. We could do more checking to say you can have up to K trailing semicolons, where K is the number of dimensions minus one.